File: ios/Domain/ProgressSnapshot.swift
import Foundation

struct ProgressSnapshot {
  let operationId: String
  let phase: String
  let processedBytes: UInt64
  let totalBytes: UInt64?
  let processedEntries: Int
  let totalEntries: Int?
  let currentEntry: String?
  let bytesPerSecond: Double?
  let estimatedSecondsRemaining: Double?
  let percentage: Double?

  init(
    operationId: String,
    phase: String = "preparing",
    processedBytes: UInt64 = 0,
    totalBytes: UInt64? = nil,
    processedEntries: Int = 0,
    totalEntries: Int? = nil,
    currentEntry: String? = nil,
    bytesPerSecond: Double? = nil,
    estimatedSecondsRemaining: Double? = nil
  ) {
    self.operationId = operationId
    self.phase = phase
    self.processedBytes = processedBytes
    self.totalBytes = totalBytes
    self.processedEntries = processedEntries
    self.totalEntries = totalEntries
    self.currentEntry = currentEntry
    self.bytesPerSecond = bytesPerSecond
    self.estimatedSecondsRemaining = estimatedSecondsRemaining

    if let total = totalBytes, total > 0 {
      self.percentage = min(Double(processedBytes) / Double(total) * 100.0, 100.0)
    } else if let total = totalEntries, total > 0 {
      self.percentage = min(Double(processedEntries) / Double(total) * 100.0, 100.0)
    } else {
      self.percentage = nil
    }
  }
}

File: android/src/main/java/com/mcodex/nitroarchive/domain/ProgressSnapshot.kt
package com.mcodex.nitroarchive.domain

data class ProgressSnapshot(
    val operationId: String,
    val phase: String = "preparing",
    val processedBytes: Long = 0,
    val totalBytes: Long? = null,
    val processedEntries: Int = 0,
    val totalEntries: Int? = null,
    val currentEntry: String? = null,
    val bytesPerSecond: Double? = null,
    val estimatedSecondsRemaining: Double? = null
) {
    val percentage: Double?
        get() {
            if (totalBytes != null && totalBytes > 0) {
                return minOf(processedBytes.toDouble() / totalBytes.toDouble() * 100.0, 100.0)
            }
            if (totalEntries != null && totalEntries > 0) {
                return minOf(processedEntries.toDouble() / totalEntries.toDouble() * 100.0, 100.0)
            }
            return null
        }
}
